lib/grammers-client/src/client/auth.rs

```diff
@@ -258,13 +270,129 @@ impl Client {
                     terms_of_service: x.terms_of_service.map(|tos| tos.into()),
                 })
             }
+            Err(InvocationError::Rpc(RpcError { name, .. }))
+                if name == "SESSION_PASSWORD_NEEDED" =>
+            {
+                let password_token = self.get_password_information().await;
+                match password_token {
+                    Ok(token) => Err(SignInError::PasswordRequired(token)),
+                    Err(e) => Err(SignInError::Other(e)),
+                }
+            }
             Err(InvocationError::Rpc(RpcError { name, .. })) if name.starts_with("PHONE_CODE_") => {
                 Err(SignInError::InvalidCode)
             }
             Err(error) => Err(SignInError::Other(error)),
         }
     }
 
+    /// Extract information needed for the two-factor authentication
+    /// It's called automatically when we get SESSION_PASSWORD_NEEDED error during sign in.
+    async fn get_password_information(&mut self) -> Result<PasswordToken, InvocationError> {
```

Comment:
I was thinking, if the parameters are time-sensitive, we might want to re-fetch them inside `check_password` (because then we'll use them immediately), not twice in `sign_in` (user may take a while to input the password).

Reply:
You mean, we fetch the parameters in `sign_in`, then ask the user for a password, check the params in `check_password` and refetch (if necessary) in `check_password` as well? This might work, why not.

Reply:
However, did an experiment with a two minutes delay to enter the password and it worked fine.
TBH, I would say that it would be cool if we could verify the params in the background while the user enters his password (because it takes some time).

Reply:
> You mean, we fetch the parameters in `sign_in`, then ask the user for a password, check the params in `check_password` and refetch (if necessary) in `check_password` as well?

Yes, exactly that.

> did an experiment with a two minutes delay to enter the password and it worked fine.

Yeah but who knows when working with Telegram, perhaps it randomly fails at times to make sure clients verify it and two attempts is enough. Or perhaps Telegram will never fail and it's only to prevent malicious attackers. Or perhaps in re-fetch it gives fresh values (haven't tried) so it's okay to do anyway.

> it would be cool if we could verify the params in the background while the user enters his password

I am not concerned with speed at the moment, this would be a fair bit of complexity not worth it right now. Furthermore, if the user knows their password, there's no need to ask for input as they may hardcode it and just call the method immediately. Then they can do this with as many clients or tasks they want to make the most of their CPUs.